baybe/parameters/utils.py

```diff
@@ -0,0 +1,81 @@
+"""Parameter utilities."""
+
+from typing import Any, Callable, Collection, Dict, List, Optional, TypeVar
+
+import pandas as pd
+
+from baybe.parameters.base import Parameter
+
+_T = TypeVar("_T", bound=Parameter)
+
+
+def get_parameters_from_dataframe(
+    df: pd.DataFrame,
+    factory: Callable[[str, Collection[Any]], _T],
+    parameters: Optional[List[_T]] = None,
+) -> List[_T]:
+    """Create a list of parameters from a dataframe.
+
+    Returns one parameter for each column of the given dataframe. By default,
+    the parameters are created using the provided factory, which takes the name
+    of the column and its unique values as arguments. However, there is also
+    the possibility to provide explicit parameter objects with names matching specific
+    columns of the dataframe, to bypass the parameter factory creation for those
+    columns. This allows finer control, for example, to specify custom parameter
+    attributes (e.g. specific optional arguments) compared to what would be provided
+    by the factory. Still, the pre-specified parameters are validated to ensure that
+    they are compatible with the contents of the dataframe.
+
+    Args:
+        df: The dataframe from which to create the parameters.
+        factory: A parameter factor, creating parameter objects for the columns
+            from the column name and the unique column values.
+        parameters: An optional list of parameter objects to bypass the factory
+            creation for columns whose names match with the parameter names.
+
+    Returns:
+        The combined parameter list, containing both the (validated) pre-specified
+        parameters and the parameters inferred from the dataframe.
+
+    Raises:
+        ValueError: If several parameters with identical names are provided.
+        ValueError: If a parameter was specified for which no match was found.
+    """
+    # Turn the pre-specified parameters into a dict and check for duplicate names
+    specified_params: Dict[str, Parameter] = {}
+    if parameters is not None:
+        for param in parameters:
+            if param.name in specified_params:
+                raise ValueError(
+                    f"You provided several parameters with the name '{param.name}'."
+                )
+            specified_params[param.name] = param
+
+    # Try to find a parameter match for each dataframe column
+    parameters = []
+    for name, series in df.items():
+        assert isinstance(name, str)
```

Comment:
@Scienfitz: since you now also collected some experience already with mypy and in particular pandas, do you perhaps have a more clever way how to resolve the type ambiguity here? Had to insert the assert statements for type narrowing from `Hashable` (which is returned from `items()`) to `str`. Wouldn't now how else to make mypy aware that we expect the dataframe to hold string valued column names only, since pandas dataframes unfortunately cannot be typed. The alternative would be a `typing.cast`, which would just disable mypy's complains here, though the complains are valid!

Reply:
uff hard to follow without seeing the original error and line, can you post?

in worst case a `cast` is fine

Reply:
Well, without the cast, two methods called in the body that expect a string raise a warning. For example, the call to `factory` gives:
<img width="311" alt="image" src="https://github.com/emdgroup/baybe/assets/23265127/a086eca1-b68c-4444-bfe4-e5b414eb95fb">
So not sure what is the better approach. Cast has no runtime overhead, but assert actually checks that it's a string (which we have nowhere ensured, actually).

Reply:
only way to avoid it then is to add dataframe valdiation, whcih seems overkill to me? assert with human-understandable error comment seems more reasonable to me then

Reply:
Ok, have done so. Will resolve. From my side, the PR is ready. Once you resolve the other thread, I'll proceed to merge 🚀